Extract GeneratorExecutor system from World

dktapps opened the pull request:
This is aimed at enabling chunk generators to exist which don't use async tasks, as described in #6556.

### Related issues & PRs
#6556

## Changes
### API changes
- `AsyncGeneratorExecutor` class added that encapsulates the logic of generating chunks using async tasks as previously
- `GeneratorExecutor` interface added that can be implemented to provide chunks in other ways
- `SyncGeneratorExecutor` which invokes the generator directly on the main thread, useful for simple generators like `Flat` where async tasks are not needed
- Some redundant APIs were removed from `World` (these will probably come back as deprecated stubs for the remainder of 5.x, but I was having too much fun deleting code)
- Removed internal `World->registerGeneratorToWorker()` (no longer useful)

### Behavioural changes
- `Flat` world generator now runs on the main thread, which is faster, has less overhead, and uses less memory

## Backwards compatibility
Backwards compatible, minus some easily stubbable API method removals.
<!-- Any possible backwards incompatible changes? How are they solved, or how can they be solved? -->
<!-- If not, you can delete this section -->

## Follow-up
Allow providing a `World` with a custom `ChunkGenerator` impl? May be useful for chunk server implementations (e.g. using BDS as a generator source)
<!-- Suggest any actions to be done before/after merging this pull request -->
<!-- For example, future changes that this PR lays the groundwork for -->

## Tests
Manually tested with `normal` and `flat` generators

dktapps committed:
First look at separating main-thread chunk gen impl from World

this is intended to open the gates to having generators that run on the main thread (e.g. flat), or that use other methods of generating chunks entirely (such as requesting them from a chunk server).

dktapps committed:
Run Flat world generators on the main thread

dktapps committed:
Yes, I know...
it can't be helped for now

dktapps committed:
Tidy, allow using closures to setup generators instead of class strings

dktapps committed:
quiet

ShockedPlot7560 reviewed (dismissed):
As it look to a lot of cut code from World to AsyncGenerator, it look OK to me. I didn't analyse the logic in its entirety (I compared each function, seeing that the vast majority of them were copied and pasted)

I like the fact that it's fully BC compatible with current registered generators

You probably had a lot of fun deleting so much code from World 😂

dktapps commented:
Couple of issues remain:

- I don't like the way I avoided the cycle between World and GeneratorExecutor
- PopulationTask's invocation, chunk unlocking etc. is duplicated between the two executor types. However the code is dissimilar enough that I'm not sure how to generalize it.
- `World->getId()` becomes conditionally useless with this change. We ought to use some ID specifically designated by the `AsyncGeneratorExecutor`, since that's the only place that actually uses it. We'd be able to deprecate numeric world IDs once this is done.

dktapps commented:
This might also be useful for generating chunks when plugins need to place objects that aren't thread safe, such as tiles. In such cases it might also be desirable to decide whether the individual stages of generation should be async or not. For example it might be useful to have base terrain generation run asynchronously and have (some) populators run on the main thread. The current architecture isn't flexible enough for this, but it would be nice if an easy way to do this were to pop out of the woodwork.

dktapps commented:
Also need to take a closer look at the sync generator to make sure it sets the dirty flags properly on new chunks, otherwise new chunks won't be saved correctly

dktapps committed:
Fixed async generator context not getting unregistered on shutdown

dktapps committed:
Move a bunch of crap into the executor namespace

dktapps committed:
Merge branch 'minor-next' into chunk-generator-extract

dktapps committed:
uhhh maybe we don't want to repeatedly populate the same chunk forever???

dktapps committed:
AsyncGeneratorExecutor: remove nonsensical code
if we're going through this handler, the only possible reason is that the chunk was populated.
We *always* want to call a population event in this case, regardless of the state of the old
chunk, and we fully expect the new chunk to always be populated in this case.

dktapps committed:
Remove unnecessary chunk loader registration
we already do this in the following loop anyway.

dktapps committed:
AsyncGeneratorExecutor: shorten a bunch of field and function names
these don't need to be so verbose now they're encapsulated.

dktapps committed:
Consistently use internal logger field

dktapps committed:
Rename field

dktapps committed:
Move generator context ID key to AsyncGeneratorExecutor
this clears the way for getting rid of numeric world IDs, which no one actually uses anyway.

dktapps committed:
um actually I think we need this

dktapps committed:
Reduce common logic duplication

dktapps committed:
Make sure we don't break people using PopulationTask

dktapps committed:
Merge branch 'minor-next' into chunk-generator-extract

dktapps committed:
We need to distinguish between removing from backlog and cancelling active tasks

dktapps committed:
ended up not needing this change

dktapps commented:
Closing in favour of the (greatly simplified) #6682
